File: Individual Project/CT07-PeakFreq-Validation/results/ct07/llm-validation/evidence/candidates/sonnet_julia/original_generation.jl
# ae_analysis_ct07.jl
#
# Acoustic Emission (AE) Analysis – CT-07 Commercial Tensile Test
# ────────────────────────────────────────────────────────────────
# Julia translation of the original MATLAB AE analysis framework.
#
# Dependencies — install once in the Julia REPL before running:
#
#   using Pkg
#   Pkg.add(["XLSX", "DataFrames", "Plots"])
#
# Usage:
#   julia ae_analysis_ct07.jl
#
# Julia-specific notes
# ───────────────────
#   • Column indexing is 1-based in both Julia and MATLAB, so df[:, N]
#     maps directly to MATLAB's data(:, N).
#   • XLSX.readtable treats row 1 as column headers and returns the rest as
#     data, matching MATLAB's xlsread behaviour for numeric-only extraction.
#   • Broadcasting dot syntax (e.g. `isfinite.(v)`) is used for
#     element-wise operations; the comprehension in `to_float` is an explicit
#     form of the same pattern, chosen for clarity when mixing type checks.
#   • `twinx(p)` from Plots.jl overlays a second y-axis sharing p's x-axis,
#     mirroring MATLAB's `yyaxis right`. Plotting to the returned subplot
#     reference controls the right axis; displaying `p` renders both axes.

using XLSX
using DataFrames
using Plots
gr()  # GR backend: fast rendering, no external window dependencies

# ═══════════════════════════════════════════════════════════════════════════════
# § 1  Data Loading
# ═══════════════════════════════════════════════════════════════════════════════

"""
    load_ae_data(filepath, sheet_name) -> NamedTuple

Load AE tensile-test data from *filepath* (Excel workbook), reading
*sheet_name*, and return a NamedTuple of plain `Float64` vectors.

Column mapping — 1-based indices match the original MATLAB `xlsread` call:

| Index | Signal    | Units |
|-------|-----------|-------|
|  1    | t1        | s     |  ← time axis for stress signal
|  3    | strain    | %     |
|  4    | stress    | MPa   |
|  5    | time      | s     |  ← time axis for AE signals
|  9    | rms       | a.u.  |  normalised RMS amplitude
| 10    | cumrms    | a.u.  |  normalised cumulative RMS
| 12    | energy    | a.u.  |  normalised AE energy
| 14    | cumenergy | a.u.  |  normalised cumulative AE energy

Empty / missing cells are replaced with `NaN` so Plots.jl skips them
silently rather than throwing a conversion error.
"""
function load_ae_data(filepath::String, sheet_name::String)

    # XLSX.readtable returns a Tables.jl-compatible object.
    # Wrapping in DataFrame gives us clean positional column access (df[:, N]).
    df = DataFrame(XLSX.readtable(filepath, sheet_name))

    # Element-wise type coercion with safe handling of both `missing`
    # (XLSX.jl ≥ 0.10 default) and `nothing` (older API, empty cells).
    # This is the broadcasting-equivalent manual form.
    function to_float(col)::Vector{Float64}
        return [ismissing(v) || isnothing(v) ? NaN : Float64(v) for v in col]
    end

    return (;
        t1        = to_float(df[:, 1]),   # stress signal time axis
        strain    = to_float(df[:, 3]),   # engineering strain (%)
        stress    = to_float(df[:, 4]),   # engineering stress (MPa)
        time      = to_float(df[:, 5]),   # AE signal time axis
        rms       = to_float(df[:, 9]),   # normalised RMS
        cumrms    = to_float(df[:, 10]),  # normalised cumulative RMS
        energy    = to_float(df[:, 12]),  # normalised AE energy
        cumenergy = to_float(df[:, 14]),  # normalised cumulative AE energy
    )
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 2  Shared Helper – Stress vs. Time (left y-axis base)
# ═══════════════════════════════════════════════════════════════════════════════

"""
    _stress_time_base(t1, stress; xlim_max, fig_title) -> Plots.Plot

Internal helper: build the blue Stress–Time left-axis plot.

Factored out because Figures 2–5 all share an identical left axis before
overlaying different AE signals on the right axis via `twinx()`.

`filter(isfinite, stress)` excludes both `NaN` and `Inf` values introduced
by `to_float` when computing the dynamic y-axis upper limit.
"""
function _stress_time_base(
        t1::Vector{Float64},
        stress::Vector{Float64};
        xlim_max::Real,
        fig_title::String)

    max_stress = maximum(filter(isfinite, stress))

    return plot(t1, stress;
        color     = :blue,
        linewidth = 1.5,
        xlabel    = "Time (s)",
        ylabel    = "Stress (MPa)",
        ylims     = (0.0, max_stress),
        xlims     = (0.0, Float64(xlim_max)),
        title     = fig_title,
        minorgrid = true,   # Julia/Plots equivalent of MATLAB's grid('minor')
        legend    = false)
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 3  Figure 1 – Strain vs. Stress
# ═══════════════════════════════════════════════════════════════════════════════

"""
    fig1_strain_stress(strain, stress) -> Plots.Plot

Single-axis line plot: Engineering Strain (%) on x, Stress (MPa) on y.
Mirrors MATLAB Figure 1.
"""
function fig1_strain_stress(strain::Vector{Float64}, stress::Vector{Float64})
    return plot(strain, stress;
        color     = :blue,
        linewidth = 1.5,
        xlabel    = "Strain (%)",
        ylabel    = "Stress (MPa)",
        ylims     = (0, 520),
        title     = "T07 Strain vs. Stress",
        minorgrid = true,
        legend    = false)
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 4  Figure 2 – Normalised RMS Profile
# ═══════════════════════════════════════════════════════════════════════════════

"""
    fig2_rms_profile(t1, stress, time, rms) -> Plots.Plot

Dual-axis plot:
  Left  (blue)  — Stress vs. Time
  Right (red)   — Normalised RMS vs. Time

⚠ The title "T09" is preserved verbatim from the original MATLAB script;
it is almost certainly a copy-paste error (should be "T07").
"""
function fig2_rms_profile(
        t1::Vector{Float64}, stress::Vector{Float64},
        time::Vector{Float64}, rms::Vector{Float64})

    # Left axis (mirrors MATLAB's `yyaxis left`)
    p = _stress_time_base(t1, stress;
            xlim_max  = 250,
            fig_title = "T09 Commercial Normalised RMS Profile")  # 'T09': as per original MATLAB

    # Right axis (mirrors MATLAB's `yyaxis right`).
    # twinx(p) adds an overlapping subplot sharing p's x-axis and returns a
    # reference to it.  Attributes set here govern the right y-axis only.
    p_right = twinx(p)
    plot!(p_right, time, rms;
        color     = :red,
        linewidth = 1.5,
        ylabel    = "Normalised RMS (a.u.)",
        ylims     = (0, 1.05),
        legend    = false)

    return p  # `p` now contains both axes; display(p) renders the full figure
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 5  Figure 3 – Normalised Cumulative RMS
# ═══════════════════════════════════════════════════════════════════════════════

"""
    fig3_cumrms_profile(t1, stress, time, cumrms) -> Plots.Plot

Dual-axis plot:
  Left  (blue) — Stress vs. Time
  Right (red)  — Normalised Cumulative RMS vs. Time
"""
function fig3_cumrms_profile(
        t1::Vector{Float64}, stress::Vector{Float64},
        time::Vector{Float64}, cumrms::Vector{Float64})

    p = _stress_time_base(t1, stress;
            xlim_max  = 300,
            fig_title = "T07 Commercial Normalised Cumulative RMS")

    p_right = twinx(p)
    plot!(p_right, time, cumrms;
        color     = :red,
        linewidth = 1.5,
        ylabel    = "Normalised Cumulative RMS (a.u.)",
        ylims     = (0, 1.05),
        legend    = false)

    return p
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 6  Figure 4 – Normalised AE Energy (scatter)
# ═══════════════════════════════════════════════════════════════════════════════

"""
    fig4_energy_profile(t1, stress, time, energy) -> Plots.Plot

Dual-axis plot:
  Left  (blue) — Stress vs. Time (line)
  Right (red)  — Normalised AE Energy vs. Time (scatter)

Uses `scatter!` to replicate MATLAB's `scatter()` call.
`markerstrokewidth = 0` removes the border ring on each marker
(equivalent to MATLAB's `markerstrokewidth = 0`).
`markersize = 5` matches MATLAB's `markersize = 5`.
"""
function fig4_energy_profile(
        t1::Vector{Float64}, stress::Vector{Float64},
        time::Vector{Float64}, energy::Vector{Float64})

    p = _stress_time_base(t1, stress;
            xlim_max  = 300,
            fig_title = "T07 Commercial Normalised AE Energy")

    p_right = twinx(p)
    scatter!(p_right, time, energy;
        markersize        = 5,
        markerstrokewidth = 0,      # no border ring — matches MATLAB original
        color             = :red,
        ylabel            = "Normalised AE Energy (a.u.)",
        ylims             = (0, 1.05),
        legend            = false)

    return p
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 7  Figure 5 – Normalised Cumulative AE Energy
# ═══════════════════════════════════════════════════════════════════════════════

"""
    fig5_cumenergy_profile(t1, stress, time, cumenergy) -> Plots.Plot

Dual-axis plot:
  Left  (blue) — Stress vs. Time
  Right (red)  — Normalised Cumulative AE Energy vs. Time
"""
function fig5_cumenergy_profile(
        t1::Vector{Float64}, stress::Vector{Float64},
        time::Vector{Float64}, cumenergy::Vector{Float64})

    p = _stress_time_base(t1, stress;
            xlim_max  = 300,
            fig_title = "T07 Commercial Normalised Cumulative AE Energy")

    p_right = twinx(p)
    plot!(p_right, time, cumenergy;
        color     = :red,
        linewidth = 1.5,
        ylabel    = "Normalised Cumulative AE Energy (a.u.)",
        ylims     = (0, 1.05),
        legend    = false)

    return p
end

# ═══════════════════════════════════════════════════════════════════════════════
# § 8  Entry Point
# ═══════════════════════════════════════════════════════════════════════════════

"""
    main() -> NTuple{5, Plots.Plot}

Orchestrates data loading and figure generation for the CT-07 AE analysis.
Returns all five plot objects so the caller can optionally save them:

    figs = main()
    savefig(figs[1], "ct07_strain_stress.png")
"""
function main()
    filepath   = "Tensile-processed.xlsx"
    sheet_name = "CT-07"

    println("─── CT-07 AE Analysis ─────────────────────────────────────")
    println("  File  : $filepath")
    println("  Sheet : $sheet_name")

    d = load_ae_data(filepath, sheet_name)
    println("  Loaded $(length(d.t1)) data points.\n")

    # Build figures
    fig1 = fig1_strain_stress(d.strain, d.stress)
    fig2 = fig2_rms_profile(d.t1, d.stress, d.time, d.rms)
    fig3 = fig3_cumrms_profile(d.t1, d.stress, d.time, d.cumrms)
    fig4 = fig4_energy_profile(d.t1, d.stress, d.time, d.energy)
    fig5 = fig5_cumenergy_profile(d.t1, d.stress, d.time, d.cumenergy)

    # Render — each `display` call opens a GR window (or inlines in Jupyter)
    display(fig1)
    display(fig2)
    display(fig3)
    display(fig4)
    display(fig5)

    println("All five figures rendered successfully.")
    return (fig1, fig2, fig3, fig4, fig5)
end

# Run when executed as a script: `julia ae_analysis_ct07.jl`
main()

File: Individual Project/CT07-PeakFreq-Validation/results/ct07/llm-validation/evidence/candidates/sonnet_python/original_generation.py
"""
ae_analysis.py
==============
Python translation of the MATLAB Acoustic Emission (AE) analysis script.

Key translation decisions:
  - xlsread()        →  pandas.read_excel()  (preserves dtypes, handles headers)
  - MATLAB 1-based   →  Python 0-based column indexing via .iloc
  - yyaxis left/right →  matplotlib twinx() with colour-coded spines/ticks
  - grid('minor')    →  ax.minorticks_on() + ax.grid(which='both')
  - scatter(…,5,…)   →  ax.scatter(…, s=5) where s is marker area in pt²

Dependencies:  pandas, openpyxl, matplotlib
Install via:   pip install pandas openpyxl matplotlib
"""

import pandas as pd
import matplotlib.pyplot as plt


# ---------------------------------------------------------------------------
# 1. Data Loading
# ---------------------------------------------------------------------------
# MATLAB's xlsread() silently skips text rows and returns numeric data only.
# pandas read_excel() keeps the full sheet.
#   • header=0  →  treat the first row as column names  (most common case)
#   • header=None →  use if the sheet has NO header row (all rows are data)
# Adjust the file extension to match your actual file (.xlsx / .xls).
EXCEL_FILE = "Commercial Tensile Tests.xlsx"
SHEET_NAME = "CT07"

df = pd.read_excel(EXCEL_FILE, sheet_name=SHEET_NAME, header=0)


# ---------------------------------------------------------------------------
# 2. Column Extraction  (MATLAB 1-based → Python 0-based via .iloc)
# ---------------------------------------------------------------------------
#  MATLAB col  │  .iloc index  │  Variable    │  Description
#  ──────────────────────────────────────────────────────────
#      1        │      0        │  t1          │  Mechanical time axis
#      3        │      2        │  strain      │  Strain (%)
#      4        │      3        │  stress      │  Stress (MPa)
#      5        │      4        │  time        │  AE time axis
#      9        │      8        │  rms         │  Normalised RMS
#     10        │      9        │  cumrms      │  Normalised Cumulative RMS
#     12        │     11        │  energy      │  Normalised AE Energy
#     14        │     13        │  cumenergy   │  Normalised Cumulative AE Energy
t1        = df.iloc[:, 0]
strain    = df.iloc[:, 2]
stress    = df.iloc[:, 3]
time      = df.iloc[:, 4]
rms       = df.iloc[:, 8]
cumrms    = df.iloc[:, 9]
energy    = df.iloc[:, 11]
cumenergy = df.iloc[:, 13]


# ---------------------------------------------------------------------------
# 3. Reusable Helper Functions
# ---------------------------------------------------------------------------

def apply_minor_grid(ax: plt.Axes) -> None:
    """
    Replicate MATLAB's grid('minor'): enable both major and minor grid lines.
    Applied to the *primary* (left) axis only; the twinx right axis is
    transparent by default, so the same grid is visible behind both datasets.
    """
    ax.minorticks_on()
    ax.grid(which="major", linestyle="-",  linewidth=0.6, alpha=0.7, zorder=0)
    ax.grid(which="minor", linestyle=":",  linewidth=0.4, alpha=0.5, zorder=0)


def plot_stress_left(
    ax: plt.Axes,
    t1_data: pd.Series,
    stress_data: pd.Series,
    xlim: float,
) -> None:
    """
    Draw the stress-vs-time curve on the left y-axis (blue).

    Colours the spine, label, and ticks blue to mirror MATLAB's
    yyaxis left visual convention.

    Parameters
    ----------
    ax          : primary (left) matplotlib Axes
    t1_data     : mechanical time series
    stress_data : stress series (MPa)
    xlim        : upper x-axis limit (seconds)
    """
    ax.plot(t1_data, stress_data, color="blue", linewidth=1.2)
    ax.set_xlabel("Time (s)")
    ax.set_ylabel("Stress (MPa)", color="blue")
    ax.set_ylim(0, stress_data.max())
    ax.set_xlim(0, xlim)
    # Colour axis elements blue to match MATLAB's yyaxis left behaviour
    ax.tick_params(axis="y", colors="blue")
    ax.spines["left"].set_color("blue")


def style_right_axis(ax: plt.Axes, ylabel: str, color: str = "red") -> None:
    """
    Apply shared styling to the right y-axis created by twinx().

    All normalised AE quantities share the [0, 1.05] y-range.
    Colours the spine, label, and ticks red to mirror MATLAB's
    yyaxis right visual convention.

    Parameters
    ----------
    ax     : secondary (right) matplotlib Axes from twinx()
    ylabel : y-axis label string
    color  : axis / label colour (default 'red')
    """
    ax.set_ylabel(ylabel, color=color)
    ax.set_ylim(0, 1.05)
    ax.tick_params(axis="y", colors=color)
    ax.spines["right"].set_color(color)


# ---------------------------------------------------------------------------
# 4. Figure 1 – Strain vs Stress
# ---------------------------------------------------------------------------
fig1, ax1 = plt.subplots(figsize=(7, 5))

ax1.plot(strain, stress, color="blue", linewidth=1.2)
ax1.set_xlabel("Strain (%)")
ax1.set_ylabel("Stress (MPa)")
ax1.set_ylim(0, 520)
ax1.set_title("T07 Strain vs. Stress")
apply_minor_grid(ax1)

fig1.tight_layout()


# ---------------------------------------------------------------------------
# 5. Figure 2 – Stress (left) + Normalised RMS (right) vs Time
#
#    NOTE: The title 'T09' in the original MATLAB script appears to be a
#    copy-paste inconsistency (the sheet is CT07). Preserved here as-is.
# ---------------------------------------------------------------------------
fig2, ax2l = plt.subplots(figsize=(8, 5))
plot_stress_left(ax2l, t1, stress, xlim=250)

# twinx() creates a new Axes sharing the same x-axis → yyaxis right equivalent
ax2r = ax2l.twinx()
ax2r.plot(time, rms, color="red", linewidth=1.0)
style_right_axis(ax2r, ylabel="Normalised RMS (a.u.)")

ax2l.set_title("T09 Commercial Normalised RMS Profile")
apply_minor_grid(ax2l)

fig2.tight_layout()


# ---------------------------------------------------------------------------
# 6. Figure 3 – Stress (left) + Normalised Cumulative RMS (right) vs Time
# ---------------------------------------------------------------------------
fig3, ax3l = plt.subplots(figsize=(8, 5))
plot_stress_left(ax3l, t1, stress, xlim=300)

ax3r = ax3l.twinx()
ax3r.plot(time, cumrms, color="red", linewidth=1.0)
style_right_axis(ax3r, ylabel="Normalised Cumulative RMS (a.u.)")

ax3l.set_title("T07 Commercial Normalised Cumulative RMS")
apply_minor_grid(ax3l)

fig3.tight_layout()


# ---------------------------------------------------------------------------
# 7. Figure 4 – Stress (left) + Normalised AE Energy scatter (right) vs Time
#
#    MATLAB: scatter(time, energy, 5, 'filled', 'r')
#    Python:  ax.scatter(time, energy, s=5, color='red')
#      • s=5  → marker area in points² (matches MATLAB's size argument)
#      • matplotlib scatter markers are filled by default (no extra kwarg needed)
#      • alpha=0.7 improves readability for dense AE datasets
# ---------------------------------------------------------------------------
fig4, ax4l = plt.subplots(figsize=(8, 5))
plot_stress_left(ax4l, t1, stress, xlim=300)

ax4r = ax4l.twinx()
ax4r.scatter(
    time, energy,
    s=5,
    color="red",
    marker="o",
    alpha=0.7,
)
style_right_axis(ax4r, ylabel="Normalised AE Energy (a.u.)")

ax4l.set_title("T07 Commercial Normalised AE Energy")
apply_minor_grid(ax4l)

fig4.tight_layout()


# ---------------------------------------------------------------------------
# 8. Figure 5 – Stress (left) + Normalised Cumulative AE Energy (right) vs Time
# ---------------------------------------------------------------------------
fig5, ax5l = plt.subplots(figsize=(8, 5))
plot_stress_left(ax5l, t1, stress, xlim=300)

ax5r = ax5l.twinx()
ax5r.plot(time, cumenergy, color="red", linewidth=1.0)
style_right_axis(ax5r, ylabel="Normalised Cumulative AE Energy (a.u.)")

ax5l.set_title("T07 Commercial Normalised Cumulative AE Energy")
apply_minor_grid(ax5l)

fig5.tight_layout()


# ---------------------------------------------------------------------------
# 9. Render all figures
# ---------------------------------------------------------------------------
# Call plt.show() once at the end; matplotlib renders all open figures together.
# In a Jupyter notebook, individual figures appear inline automatically and
# this call can be omitted.
plt.show()
